src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm.dataset
+
+import com.microsoft.ml.spark.lightgbm.ColumnParams
+import com.microsoft.ml.spark.lightgbm.swig.{BaseSwigArray, ChunkedArray, DoubleChunkedArray, DoubleSwigArray,
+  FloatChunkedArray, FloatSwigArray, IntChunkedArray, IntSwigArray}
+import org.apache.spark.sql.Row
+
+import scala.collection.mutable.ListBuffer
+
+object ChunkedArrayUtils {
+  def copyChunkedArray[T](chunkedArray: ChunkedArray[T],
+                          mainArray: BaseSwigArray[T],
+                          threadRowStartIndex: Long,
+                          chunkSize: Long): Unit = {
+    // Copy in parallel on each thread
+    // First copy full chunks
+    val chunkCount = chunkedArray.getChunksCount() - 1
+    for (chunk <- 0L until chunkCount) {
+      for (in_chunk_idx <- 0L until chunkSize) {
+        mainArray.setItem(threadRowStartIndex + chunk * chunkSize + in_chunk_idx,
+          chunkedArray.getItem(chunk, in_chunk_idx, (-1).asInstanceOf[T]))
+      }
+    }
+    // Next copy filled values from last chunk only
+    val lastChunkCount = chunkedArray.getLastChunkAddCount()
+    for (last_chunk_idx <- 0L until lastChunkCount) {
+      mainArray.setItem(threadRowStartIndex + chunkCount * chunkSize + last_chunk_idx,
+        chunkedArray.getItem(chunkCount, last_chunk_idx, (-1).asInstanceOf[T]))
+    }
```

Comment:
Yes, it would be better to move the code fully into the native side, so the chunked arrays would be merged to main array in c++ only, but that would require changing lightgbm code.  I can do that in a separate step/PR later on outside of mmlspark.  Right now I need to get this optimization working.  In terms of performance, this isn't a huge bottleneck right now, the swig calls seem to be fast.

Reply:
Really, well, that's a pretty happy surprise. There used to be a pretty large overhead crossing out of Java land. Just out of curiosity, do you know how fast?

Reply:
For the 9 GB regression dataset we tested on a small spark cluster with just 4 nodes all of the data preprocessing seemed to take a minute or so based on the logs compared to the training code which took the vast majority of the time (~20 mins) for 1000 iterations, and most of the data processing time is actually in getting the values from the spark dataframe which executes previous spark steps because of lazy evaluation.  On the 37 GB dataset the results were similar.  On a different 4 GB Higgs classification dataset with a spark cluster that had 8 workers the preprocessing time was just a couple seconds.


<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
</head>

<body link="#0563C1" vlink="#954F72">



  |   |   |   |   |   | Total Training | Total Training | Data Prep | LGBM
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
Dataset (GB) | Cluster | Mode | Cores per Node | Workers | Status | Time (min) | Time (h) | Time (h) | Time (h)
9 | 04Cores14GBx5 | Singleton Dataset | 4 | 4 | Succeeded | 26.57 | 0.44 | 0.08 | 0.38
9 | 08Cores28GBx5 | Singleton Dataset | 8 | 4 | Succeeded | 16.29 | 0.27 | 0.05 | 0.24
9 | 16Cores64GBx5 | Singleton Dataset | 16 | 4 | Succeeded | 18.41 | 0.31 | 0.05 | 0.28
9 | 32Cores128GBx5 | Singleton Dataset | 32 | 4 | Succeeded | 17.75 | 0.30 | 0.03 | 0.28
37 | 04Cores14GBx5 | Singleton Dataset | 4 | 4 | Failed | 9.18 | 0.15 | - | -
37 | 08Cores28GBx5 | Singleton Dataset | 8 | 4 | Failed | 7.07 | 0.12 | - | -
37 | 16Cores64GBx5 | Singleton Dataset | 16 | 4 | Succeeded | 112.2 | 1.87 | ? | ?
37 | 32Cores128GBx5 | Singleton Dataset | 32 | 4 | Succeeded | 44.37 | 0.74 | 0.05 | 0.69



</body>

</html>
</div><!--EndFragment-->
</body>
</html>